src/cpp-common/include/cpp-common/HashMap.tpp

```diff
@@ -0,0 +1,145 @@
+            if (pair.first == key) {
+                pair.second.~MappedType();
+                m_usedSpacesFlag[index] = false;
+                return true;
+            }
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return {};
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+void HashMap<Key, MappedType, maxSize>::clear() {
+    for (int i = 0; i < maxSize; i++) {
+        // Could forego calling constructor since using placement new
+        reinterpret_cast<std::pair<Key, MappedType>*>(&m_storage[i])->second.~MappedType();
+        m_usedSpacesFlag[i] = false;
+    }
+    m_usedSpaces = 0;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::get(Key k, MappedType& item) const {
+    uint16_t index = hash(k);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
+            if (pair.first == k) {
+                // emplace on reference passed
+                new (&item) MappedType(pair.second);
+                return true;
+            }
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return false;
+}
+template <typename Key, typename MappedType, uint16_t maxSize>
+std::optional<MappedType> HashMap<Key, MappedType, maxSize>::at(Key key) const {
+    uint16_t index = hash(key);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
+            if (pair.first == key) {
+                return pair.second;
+            }
+        }
+
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return {};
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+uint16_t HashMap<Key, MappedType, maxSize>::getMaxSize() const {
+    return maxSize;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::isEmpty() const {
+    return m_usedSpaces == 0;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::isFull() const {
+    return m_usedSpaces == maxSize;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+uint16_t HashMap<Key, MappedType, maxSize>::hash(Key key) {
+    return key % maxSize;
+}
+
```

Comment:
could we get a function to get the free space and used space?
like:
getFreeSize
getLength/getUsedSize